pkg/controller/shared_state.go

```diff
@@ -1345,13 +1371,171 @@ func (c *SharedState) createReplicas(ctx context.Context, remainingReplicas int,
+func (c *SharedState) unpersistAttachmentFailure(volumeName string) {
+	trySend(c.eventsToUnpersistQueue, volumeName) // also mitigates unlikely race condition of sending to closed channel
+}
+
+func (c *SharedState) _eventRefresherInitFunc() {
+	c.eventsToPersistQueue = make(chan ReplicaAttachmentFailureInfo, maxPersistEventConcurrency)
+	c.eventsToUnpersistQueue = make(chan string, maxUnpersistEventConcurrency)
+	go c._eventRefresherRoutine()
+}
+
+func (c *SharedState) _eventRefresherRoutine() {
+	const eventLifeSpan = time.Hour // ToDo: query k8s config?
+	type Event struct {
+		prev    *Event
+		next    *Event
+		message string
+		delay   time.Duration
+		objects []runtime.Object
+	}
+	eventMap := map[string]*Event{}
+	newFailureInfo := <-c.eventsToPersistQueue
+	klog.V(2).Infof("Starting the event refresher") // should this be logged via workflow instead?
+	var currEvent *Event
+	var alarm *time.Timer
+	var lastTime time.Time         // the time at which the previous event was inserted / refreshed
+	var timeVariance time.Duration // shifts the time at which every event is refreshed to mitigate the alarm's inaccuracy while maintainting the sum of all delays (as the event life span) to prevent drift
+	for {
+		// add the new failure
+		currTime := time.Now()
+		newEvent := &Event{
+			message: newFailureInfo.message,
+			objects: newFailureInfo.pods,
+		}
+		newEventTimeVariance := currTime.Sub(newFailureInfo.timestamp) + maxEventOverlapDuration
+		if currEvent == nil {
+			currEvent = newEvent
+			newEvent.delay = eventLifeSpan
+			timeVariance = newEventTimeVariance
+			alarm = time.NewTimer(eventLifeSpan - timeVariance)
+			lastTime = currTime
+		} else {
+			currEvent.prev.next = newEvent
+
+			newEvent.delay = newFailureInfo.timestamp.Sub(lastTime) + (timeVariance - newEventTimeVariance)
+			// the new event's delay is relative to the last event inserted / refreshed,
+			// because a non-empty circular linked list takes a total of 1 hour to traverse
+			// add the global time variance back into the new event's delay because its subtracted from each delay
+			// and subtract the new event's local time variance from its delay because we want to use its local time variance without affecting the global one
+
+			currEvent.delay -= newEvent.delay
+			lastTime = lastTime.Add(newEvent.delay) // the sum of lastTime and currentEvent.delay should remain the same
+			// the values of current event delay and last time are now both relative to the new event
+		}
+		newEvent.next = currEvent
+		currEvent.prev = newEvent
+
+		oldEvent := eventMap[newFailureInfo.volumeName]
+		if oldEvent != nil {
+			// remove the old event from the circular linked list
+			oldEvent.prev.next = oldEvent.next
+			oldEvent.next.prev = oldEvent.prev
+			oldEvent.next.delay += oldEvent.delay
+		}
+		eventMap[newFailureInfo.volumeName] = newEvent
+
+	NoNewFailureLoop:
+		for {
+			select {
+			case newFailureInfo = <-c.eventsToPersistQueue:
```

Comment:
Wouldn't this risk (although with a very low likelihood) not being able to refresh current event (HEAD of the linked list) at the right time if event gets queued when timer is about to go off? Just wondering because I see that you have added a lot of detail to ensure events get refreshed without drift.

Reply:
Yeah, if a new event is seen by this loop, it would add that new failure and go back into the no new failure loop; if multiple channels happen to be ready at the same time there should be no loss of events & in the scenario you describe, the alarm would/should stay in its triggered state & a delay due to consuming events would be seen by the alarm handler the same as a delay due to sleep inaccuracy.